File: loop-app/scraper/cleanup_corrupt_events.py
#!/usr/bin/env python3
"""
cleanup_corrupt_events.py
One-off, idempotent data cleanup script for LOOP Firestore events.

Detects and fixes/archives:
1. Events corrupted with year 2027 due to the previous scraper year-rollover bug.
2. Duplicate events with identical normalized (host, title).
3. Events with invalid date markers ("not available", "ongoing").
4. Events missing required cover posters.

Usage:
  python cleanup_corrupt_events.py          # Dry-run (read-only audit)
  python cleanup_corrupt_events.py --commit # Apply changes to Firestore
"""

import os
import sys
import re
from datetime import datetime
import firebase_admin
from firebase_admin import credentials, firestore
from dotenv import load_dotenv

load_dotenv()

# Initialize Firebase
CRED_PATH = os.getenv("FIREBASE_CREDENTIALS_PATH", os.path.join(os.path.dirname(__file__), "serviceAccountKey.json"))
if not os.path.exists(CRED_PATH):
    # Try parent directory
    alt_path = os.path.join(os.path.dirname(__file__), "..", "serviceAccountKey.json")
    if os.path.exists(alt_path):
        CRED_PATH = alt_path

if not firebase_admin._apps:
    try:
        cred = credentials.Certificate(CRED_PATH)
        firebase_admin.initialize_app(cred)
    except Exception as e:
        print(f"[Error] Failed to initialize Firebase Admin SDK: {e}")
        print("Please check FIREBASE_CREDENTIALS_PATH or serviceAccountKey.json.")
        sys.exit(1)

db = firestore.client()

def run_cleanup(commit: bool = False):
    print("=" * 60)
    print(f"LOOP FIRESTORE DATA CLEANUP — Mode: {'COMMIT (WRITING)' if commit else 'DRY RUN (READ ONLY)'}")
    print("=" * 60)

    events_ref = db.collection("events")
    docs = list(events_ref.stream())
    print(f"Fetched {len(docs)} total event documents from Firestore.\n")

    current_year = datetime.now().year
    seen_host_titles = {}
    to_update = []
    to_archive = []

    for doc in docs:
        d = doc.to_dict()
        doc_id = doc.id
        title = d.get("title", "")
        host = d.get("host", "")
        date_str = str(d.get("date", "")).strip().lower()
        starts_at = d.get("startsAt")
        image = d.get("image", "")
        status = d.get("status", "")

        norm_key = f"{host.lower()}:{re.sub(r'[^a-z0-9]', '', title.lower())}"

        # 1. Check for duplicates
        if norm_key in seen_host_titles:
            prev_id, prev_data = seen_host_titles[norm_key]
            # Keep the one that is approved or has an image; archive the duplicate
            print(f"[Duplicate Found] Doc '{doc_id}' duplicates '{prev_id}' ('{title}' by {host})")
            to_archive.append((doc_id, "duplicate_event", f"Duplicates document {prev_id}"))
            continue
        else:
            seen_host_titles[norm_key] = (doc_id, d)

        # 2. Check for missing/invalid poster image
        if not image or "res.cloudinary.com" not in image:
            print(f"[Missing Poster] Doc '{doc_id}' ('{title}') has no Cloudinary image.")
            to_archive.append((doc_id, "missing_poster", "No valid Cloudinary image"))
            continue

        # 3. Check for invalid date strings ("not available", "ongoing")
        invalid_markers = ["not available", "ongoing", "not specified", "none", "tbd", "tba"]
        if any(marker in date_str for marker in invalid_markers):
            print(f"[Invalid Date] Doc '{doc_id}' ('{title}') has invalid date '{d.get('date')}'.")
            to_archive.append((doc_id, "invalid_date", f"Date string is '{d.get('date')}'"))
            continue

        # 4. Check for 2027 year-rollover bug in startsAt
        if starts_at:
            dt = None
            if hasattr(starts_at, "to_datetime"):
                dt = starts_at.to_datetime()
            elif hasattr(starts_at, "year"):
                dt = starts_at
            
            if dt and dt.year > current_year:
                corrected_dt = dt.replace(year=current_year)
                print(f"[Year Rollover Bug] Doc '{doc_id}' ('{title}') has future year {dt.year} -> Correcting to {corrected_dt.year}.")
                to_update.append((doc_id, {"startsAt": corrected_dt}))

    print("\n" + "-" * 60)
    print(f"Summary: {len(to_update)} document(s) to fix, {len(to_archive)} document(s) to archive.")
    print("-" * 60)

    if not commit:
        print("\n[Dry Run] No changes written to Firestore. Pass --commit to apply.")
        return

    # Apply updates
    batch = db.batch()
    batch_count = 0

    for doc_id, fields in to_update:
        ref = events_ref.document(doc_id)
        batch.update(ref, fields)
        batch_count += 1

    for doc_id, reason, details in to_archive:
        ref = events_ref.document(doc_id)
        # Soft-archive by setting status: "archived" and recording cleanup info
        batch.update(ref, {
            "status": "archived",
            "archivedReason": reason,
            "archivedDetails": details,
            "archivedAt": firestore.SERVER_TIMESTAMP
        })
        batch_count += 1

    if batch_count > 0:
        batch.commit()
        print(f"\n[Committed] Successfully updated/archived {batch_count} documents in Firestore!")
    else:
        print("\n[Clean] Database is already clean. No changes necessary.")

if __name__ == "__main__":
    commit_mode = "--commit" in sys.argv
    run_cleanup(commit=commit_mode)

File: loop-app/scripts/cleanup_corrupt_events.js
/**
 * cleanup_corrupt_events.js
 * Node.js runner for LOOP Firestore cleanup using firebase-admin.
 * 
 * Usage:
 *   node cleanup_corrupt_events.js          # Dry-run
 *   node cleanup_corrupt_events.js --commit # Apply changes to live Firestore
 */

const admin = require('firebase-admin');
const path = require('path');
const fs = require('fs');

const serviceAccountPath = path.resolve(__dirname, '../scraper/serviceAccountKey.json');

if (!fs.existsSync(serviceAccountPath)) {
  console.error(`[Error] Service account key not found at ${serviceAccountPath}`);
  process.exit(1);
}

const serviceAccount = JSON.parse(fs.readFileSync(serviceAccountPath, 'utf8'));

if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount),
  });
}

const db = admin.firestore();

async function runCleanup(commit = false) {
  console.log('='.repeat(60));
  console.log(`LOOP FIRESTORE CLEANUP — Mode: ${commit ? 'COMMIT (APPLYING CHANGES)' : 'DRY-RUN (READ-ONLY)'}`);
  console.log('='.repeat(60));

  const eventsRef = db.collection('events');
  const snapshot = await eventsRef.get();
  console.log(`Fetched ${snapshot.size} total event documents from Firestore.\n`);

  const currentYear = new Date().getFullYear();
  const seenHostTitles = new Map();
  const toUpdate = [];
  const toArchive = [];

  snapshot.forEach((doc) => {
    const d = doc.data();
    const docId = doc.id;
    const title = (d.title || '').trim();
    const host = (d.host || '').trim();
    const dateStr = String(d.date || '').trim().toLowerCase();
    const startsAt = d.startsAt;
    const image = d.image || '';
    const status = d.status || '';

    // If already archived, skip
    if (status === 'archived') return;

    const normKey = `${host.toLowerCase()}:${title.toLowerCase().replace(/[^a-z0-9]/g, '')}`;

    // 1. Check for duplicates
    if (seenHostTitles.has(normKey)) {
      const prevId = seenHostTitles.get(normKey);
      console.log(`[Duplicate Found] Doc '${docId}' duplicates '${prevId}' ('${title}' by ${host})`);
      toArchive.push({
        id: docId,
        reason: 'duplicate_event',
        details: `Duplicates document ${prevId}`,
      });
      return;
    } else {
      seenHostTitles.set(normKey, docId);
    }

    // 2. Check for missing/invalid poster image
    if (!image || !image.includes('res.cloudinary.com')) {
      console.log(`[Missing Poster] Doc '${docId}' ('${title}') has no Cloudinary image.`);
      toArchive.push({
        id: docId,
        reason: 'missing_poster',
        details: 'No valid Cloudinary image URL',
      });
      return;
    }

    // 3. Check for invalid date strings
    const invalidMarkers = ['not available', 'ongoing', 'not specified', 'none', 'tbd', 'tba'];
    if (invalidMarkers.some((marker) => dateStr.includes(marker))) {
      console.log(`[Invalid Date] Doc '${docId}' ('${title}') has invalid date string '${d.date}'.`);
      toArchive.push({
        id: docId,
        reason: 'invalid_date',
        details: `Date string is '${d.date}'`,
      });
      return;
    }

    // 4. Check for 2027 year-rollover bug in startsAt
    if (startsAt) {
      let dt = null;
      if (typeof startsAt.toDate === 'function') {
        dt = startsAt.toDate();
      } else if (startsAt._seconds) {
        dt = new Date(startsAt._seconds * 1000);
      } else if (typeof startsAt === 'string') {
        dt = new Date(startsAt);
      }

      if (dt && !isNaN(dt.getTime()) && dt.getFullYear() > currentYear) {
        const corrected = new Date(dt);
        corrected.setFullYear(currentYear);
        console.log(`[Year Rollover Bug] Doc '${docId}' ('${title}') has year ${dt.getFullYear()} -> Correcting to ${corrected.getFullYear()}.`);
        toUpdate.push({
          id: docId,
          data: {
            startsAt: admin.firestore.Timestamp.fromDate(corrected),
          },
        });
      }
    }
  });

  console.log('\n' + '-'.repeat(60));
  console.log(`Summary: ${toUpdate.length} document(s) to fix year, ${toArchive.length} document(s) to archive.`);
  console.log('-'.repeat(60));

  if (!commit) {
    console.log('\n[Dry Run Complete] No changes written to Firestore. Pass --commit to apply.');
    return;
  }

  const batch = db.batch();
  let count = 0;

  for (const item of toUpdate) {
    const ref = eventsRef.doc(item.id);
    batch.update(ref, item.data);
    count++;
  }

  for (const item of toArchive) {
    const ref = eventsRef.doc(item.id);
    batch.update(ref, {
      status: 'archived',
      archivedReason: item.reason,
      archivedDetails: item.details,
      archivedAt: admin.firestore.FieldValue.serverTimestamp(),
    });
    count++;
  }

  if (count > 0) {
    await batch.commit();
    console.log(`\n[Committed] Successfully updated/archived ${count} documents in live Firestore!`);
  } else {
    console.log('\n[Clean] Database is already clean. No changes necessary.');
  }
}

const isCommit = process.argv.includes('--commit');
runCleanup(isCommit).catch((err) => {
  console.error('[Error] Cleanup failed:', err);
  process.exit(1);
});
